File: sdks/python/t402/src/t402/a2a/ap2.py
"""AP2 embedded flow types and bridge functions for t402 payments."""

from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any, Dict, List, Optional

from t402.a2a.types import A2AArtifact, A2AMessage, A2AMessagePart

from .helpers import (
    create_t402_extension,
    create_x402_extension,
)
from .types import (
    A2A_EXTENSIONS_HEADER,
    X402_A2A_EXTENSION_URI,
)

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

AP2_EXTENSION_URI = (
    "https://github.com/google-agentic-commerce/ap2/tree/v0.1"
)
X402_PAYMENT_METHOD = "https://www.x402.org/"

AP2_DATA_KEY_INTENT_MANDATE = "ap2.mandates.IntentMandate"
AP2_DATA_KEY_CART_MANDATE = "ap2.mandates.CartMandate"
AP2_DATA_KEY_PAYMENT_MANDATE = "ap2.mandates.PaymentMandate"
AP2_DATA_KEY_PAYMENT_RECEIPT = "ap2.PaymentReceipt"

# ---------------------------------------------------------------------------
# AP2 Types
# ---------------------------------------------------------------------------


@dataclass
class PaymentCurrencyAmount:
    """W3C PaymentCurrencyAmount."""

    currency: str
    value: float


@dataclass
class PaymentItem:
    """W3C PaymentItem."""

    label: str
    amount: PaymentCurrencyAmount
    pending: Optional[bool] = None


@dataclass
class AP2PaymentMethodData:
    """AP2 payment method entry."""

    supported_methods: str
    data: Optional[Dict[str, Any]] = None


@dataclass
class AP2PaymentDetailsInit:
    """AP2 payment details."""

    id: str
    display_items: List[PaymentItem]
    total: PaymentItem


@dataclass
class AP2PaymentRequest:
    """AP2 payment request."""

    method_data: List[AP2PaymentMethodData]
    details: AP2PaymentDetailsInit


@dataclass
class AP2PaymentResponse:
    """AP2 payment response."""

    request_id: str
    method_name: str
    details: Optional[Dict[str, Any]] = None


@dataclass
class IntentMandate:
    """AP2 intent mandate (pre-cart)."""

    natural_language_description: str
    user_cart_confirmation_required: bool
    intent_expiry: str
    merchants: Optional[List[str]] = None
    skus: Optional[List[str]] = None
    requires_refundability: Optional[bool] = None


@dataclass
class CartContents:
    """AP2 cart contents."""

    id: str
    user_cart_confirmation_required: bool
    payment_request: AP2PaymentRequest
    cart_expiry: str
    merchant_name: str


@dataclass
class CartMandate:
    """AP2 cart mandate."""

    contents: CartContents
    merchant_authorization: Optional[str] = None


@dataclass
class PaymentMandateContents:
    """AP2 payment mandate contents."""

    payment_mandate_id: str
    payment_details_id: str
    payment_details_total: PaymentItem
    payment_response: AP2PaymentResponse
    merchant_agent: str
    timestamp: str


@dataclass
class PaymentMandate:
    """AP2 payment mandate."""

    payment_mandate_contents: PaymentMandateContents
    user_authorization: Optional[str] = None


@dataclass
class AP2PaymentReceipt:
    """AP2 payment receipt."""

    payment_mandate_id: str
    timestamp: str
    payment_id: str
    amount: PaymentCurrencyAmount
    payment_status: Dict[str, Any]


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _to_dict(obj: Any) -> Any:
    """Convert a dataclass to a plain dict."""
    return asdict(obj)


# ---------------------------------------------------------------------------
# Bridge functions
# ---------------------------------------------------------------------------


def create_cart_mandate_with_x402(
    contents: CartContents,
    requirements: List[Dict[str, Any]],
    merchant_authorization: Optional[str] = None,
) -> CartMandate:
    """Create a CartMandate with x402 requirements embedded."""
    methods = [
        m
        for m in contents.payment_request.method_data
        if m.supported_methods != X402_PAYMENT_METHOD
    ]
    methods.append(
        AP2PaymentMethodData(
            supported_methods=X402_PAYMENT_METHOD,
            data={"requirements": requirements},
        )
    )
    updated = CartContents(
        id=contents.id,
        user_cart_confirmation_required=(
            contents.user_cart_confirmation_required
        ),
        payment_request=AP2PaymentRequest(
            method_data=methods,
            details=contents.payment_request.details,
        ),
        cart_expiry=contents.cart_expiry,
        merchant_name=contents.merchant_name,
    )
    return CartMandate(
        contents=updated,
        merchant_authorization=merchant_authorization,
    )


def extract_x402_requirements(
    mandate: CartMandate,
) -> Optional[List[Dict[str, Any]]]:
    """Extract x402 requirements from a CartMandate."""
    for m in mandate.contents.payment_request.method_data:
        if (
            m.supported_methods == X402_PAYMENT_METHOD
            and m.data
        ):
            reqs = m.data.get("requirements")
            if isinstance(reqs, list):
                return reqs
    return None


def create_payment_mandate_with_x402(
    contents: PaymentMandateContents,
    payload: Dict[str, Any],
    user_authorization: Optional[str] = None,
) -> PaymentMandate:
    """Create a PaymentMandate with x402 payload."""
    updated = PaymentMandateContents(
        payment_mandate_id=contents.payment_mandate_id,
        payment_details_id=contents.payment_details_id,
        payment_details_total=contents.payment_details_total,
        payment_response=AP2PaymentResponse(
            request_id=contents.payment_response.request_id,
            method_name=X402_PAYMENT_METHOD,
            details=payload,
        ),
        merchant_agent=contents.merchant_agent,
        timestamp=contents.timestamp,
    )
    return PaymentMandate(
        payment_mandate_contents=updated,
        user_authorization=user_authorization,
    )


def extract_x402_payload(
    mandate: PaymentMandate,
) -> Optional[Dict[str, Any]]:
    """Extract x402 payload from a PaymentMandate."""
    resp = mandate.payment_mandate_contents.payment_response
    if resp.method_name != X402_PAYMENT_METHOD:
        return None
    return resp.details


def create_ap2_extension(
    roles: Optional[List[str]] = None,
    required: bool = False,
) -> Dict[str, Any]:
    """Create an AP2 extension declaration."""
    if roles is None:
        roles = ["merchant"]
    return {
        "uri": AP2_EXTENSION_URI,
        "description": (
            f"AP2 payment agent (roles: {', '.join(roles)})."
        ),
        "required": required,
    }


# ---------------------------------------------------------------------------
# DataPart helpers
# ---------------------------------------------------------------------------


def create_cart_mandate_data_part(
    mandate: CartMandate,
) -> A2AMessagePart:
    """Create a DataPart containing a CartMandate."""
    return A2AMessagePart(
        kind="data",
        data={AP2_DATA_KEY_CART_MANDATE: _to_dict(mandate)},
    )


def create_payment_mandate_data_part(
    mandate: PaymentMandate,
) -> A2AMessagePart:
    """Create a DataPart containing a PaymentMandate."""
    return A2AMessagePart(
        kind="data",
        data={
            AP2_DATA_KEY_PAYMENT_MANDATE: _to_dict(mandate)
        },
    )


def create_intent_mandate_data_part(
    mandate: IntentMandate,
) -> A2AMessagePart:
    """Create a DataPart containing an IntentMandate."""
    return A2AMessagePart(
        kind="data",
        data={
            AP2_DATA_KEY_INTENT_MANDATE: _to_dict(mandate)
        },
    )


def create_payment_receipt_data_part(
    receipt: AP2PaymentReceipt,
) -> A2AMessagePart:
    """Create a DataPart containing a PaymentReceipt."""
    return A2AMessagePart(
        kind="data",
        data={AP2_DATA_KEY_PAYMENT_RECEIPT: _to_dict(receipt)},
    )


def extract_cart_mandate_from_artifact(
    artifact: A2AArtifact,
) -> Optional[Dict[str, Any]]:
    """Extract a CartMandate dict from an A2AArtifact."""
    if not artifact.parts:
        return None
    for part in artifact.parts:
        if (
            part.kind == "data"
            and part.data
            and AP2_DATA_KEY_CART_MANDATE in part.data
        ):
            return part.data[AP2_DATA_KEY_CART_MANDATE]
    return None


def extract_payment_mandate_from_message(
    message: A2AMessage,
) -> Optional[Dict[str, Any]]:
    """Extract a PaymentMandate dict from an A2AMessage."""
    for part in message.parts:
        if (
            part.kind == "data"
            and part.data
            and AP2_DATA_KEY_PAYMENT_MANDATE in part.data
        ):
            return part.data[AP2_DATA_KEY_PAYMENT_MANDATE]
    return None


# ---------------------------------------------------------------------------
# AgentCard extension composition & header helpers
# ---------------------------------------------------------------------------


def create_payment_extensions(
    *,
    ap2_roles: Optional[List[str]] = None,
    t402_required: bool = False,
    x402_required: bool = False,
    ap2_required: bool = False,
) -> List[Dict[str, Any]]:
    """Create payment extensions array for an AgentCard."""
    extensions: List[Dict[str, Any]] = [
        asdict(create_t402_extension(t402_required)),
        asdict(create_x402_extension(x402_required)),
    ]
    if ap2_roles:
        extensions.append(
            create_ap2_extension(ap2_roles, ap2_required)
        )
    return extensions


def get_payment_extension_headers(
    include_ap2: bool = False,
) -> Dict[str, str]:
    """Get HTTP headers for A2A payment extension activation."""
    uris = [X402_A2A_EXTENSION_URI]
    if include_ap2:
        uris.append(AP2_EXTENSION_URI)
    return {A2A_EXTENSIONS_HEADER: ", ".join(uris)}

File: sdks/typescript/packages/core/src/types/ap2.ts
/**
 * AP2 (Agent Payments Protocol) Types
 *
 * W3C Payment Request API subset + AP2 mandate types for the
 * embedded payment flow where x402 requirements/payloads are
 * wrapped inside AP2 mandates.
 *
 * @see https://github.com/google-agentic-commerce/ap2/tree/v0.1
 * @see https://github.com/google-agentic-commerce/a2a-x402/blob/main/spec/v0.2
 */

import type { PaymentRequirements, PaymentPayload } from "./payments";
import type { A2AExtension, A2ADataPart, A2AArtifact, A2AMessage } from "./a2a";
import {
  A2A_EXTENSIONS_HEADER,
  X402_A2A_EXTENSION_URI,
  createT402Extension,
  createX402Extension,
} from "./a2a";

// ============================================================================
// AP2 Constants
// ============================================================================

/** AP2 extension URI for agent card declarations */
export const AP2_EXTENSION_URI = "https://github.com/google-agentic-commerce/ap2/tree/v0.1";

/** x402 payment method identifier for AP2 PaymentMethodData */
export const X402_PAYMENT_METHOD = "https://www.x402.org/";

/** DataPart canonical keys for AP2 mandate types */
export const AP2_DATA_KEYS = {
  INTENT_MANDATE: "ap2.mandates.IntentMandate",
  CART_MANDATE: "ap2.mandates.CartMandate",
  PAYMENT_MANDATE: "ap2.mandates.PaymentMandate",
  PAYMENT_RECEIPT: "ap2.PaymentReceipt",
} as const;

// ============================================================================
// AP2 Role
// ============================================================================

export type AP2Role = "merchant" | "shopper" | "credentials-provider" | "payment-processor";

// ============================================================================
// W3C Payment Request API (subset)
// ============================================================================

export interface PaymentCurrencyAmount {
  currency: string;
  value: number;
}

export interface PaymentItem {
  label: string;
  amount: PaymentCurrencyAmount;
  pending?: boolean;
}

export interface PaymentMethodData {
  supported_methods: string;
  data?: Record<string, unknown>;
}

export interface PaymentDetailsInit {
  id: string;
  display_items: PaymentItem[];
  total: PaymentItem;
}

export interface AP2PaymentRequest {
  method_data: PaymentMethodData[];
  details: PaymentDetailsInit;
}

export interface AP2PaymentResponse {
  request_id: string;
  method_name: string;
  details?: Record<string, unknown>;
}

// ============================================================================
// AP2 Mandate Types
// ============================================================================

export interface IntentMandate {
  natural_language_description: string;
  user_cart_confirmation_required: boolean;
  merchants?: string[];
  skus?: string[];
  requires_refundability?: boolean;
  intent_expiry: string; // ISO 8601
}

export interface CartContents {
  id: string;
  user_cart_confirmation_required: boolean;
  payment_request: AP2PaymentRequest;
  cart_expiry: string; // ISO 8601
  merchant_name: string;
}

export interface CartMandate {
  contents: CartContents;
  merchant_authorization?: string;
}

export interface PaymentMandateContents {
  payment_mandate_id: string;
  payment_details_id: string;
  payment_details_total: PaymentItem;
  payment_response: AP2PaymentResponse;
  merchant_agent: string;
  timestamp: string; // ISO 8601
}

export interface PaymentMandate {
  payment_mandate_contents: PaymentMandateContents;
  user_authorization?: string;
}

export interface PaymentReceipt {
  payment_mandate_id: string;
  timestamp: string; // ISO 8601
  payment_id: string;
  amount: PaymentCurrencyAmount;
  payment_status:
    | { merchant_confirmation_id: string }
    | { error_message: string }
    | { failure_message: string };
}

// ============================================================================
// Bridge Functions — x402 ↔ AP2
// ============================================================================

/**
 * Create a CartMandate with x402 PaymentRequirements embedded in
 * the PaymentMethodData for method "https://www.x402.org/".
 *
 * @param cartContents - The base cart contents to embed requirements into
 * @param requirements - The x402 payment requirements to embed
 * @param merchantAuthorization - Optional merchant authorization token
 * @returns A CartMandate with x402 requirements in PaymentMethodData
 */
export function createCartMandateWithX402(
  cartContents: CartContents,
  requirements: PaymentRequirements[],
  merchantAuthorization?: string,
): CartMandate {
  const x402MethodData: PaymentMethodData = {
    supported_methods: X402_PAYMENT_METHOD,
    data: { requirements },
  };

  const updatedContents: CartContents = {
    ...cartContents,
    payment_request: {
      ...cartContents.payment_request,
      method_data: [
        ...cartContents.payment_request.method_data.filter(
          m => m.supported_methods !== X402_PAYMENT_METHOD,
        ),
        x402MethodData,
      ],
    },
  };

  return {
    contents: updatedContents,
    merchant_authorization: merchantAuthorization,
  };
}

/**
 * Extract x402 PaymentRequirements from a CartMandate.
 *
 * @param cartMandate - The CartMandate to extract requirements from
 * @returns The extracted x402 PaymentRequirements, or undefined if not found
 */
export function extractX402Requirements(
  cartMandate: CartMandate,
): PaymentRequirements[] | undefined {
  const methodData = cartMandate.contents.payment_request.method_data.find(
    m => m.supported_methods === X402_PAYMENT_METHOD,
  );
  if (!methodData?.data?.requirements) return undefined;
  return methodData.data.requirements as PaymentRequirements[];
}

/**
 * Create a PaymentMandate with x402 PaymentPayload embedded in
 * the PaymentResponse.details for method "https://www.x402.org/".
 *
 * @param mandateContents - The base payment mandate contents
 * @param payload - The x402 payment payload to embed
 * @param userAuthorization - Optional user authorization token
 * @returns A PaymentMandate with x402 payload in PaymentResponse details
 */
export function createPaymentMandateWithX402(
  mandateContents: PaymentMandateContents,
  payload: PaymentPayload,
  userAuthorization?: string,
): PaymentMandate {
  const updatedContents: PaymentMandateContents = {
    ...mandateContents,
    payment_response: {
      ...mandateContents.payment_response,
      method_name: X402_PAYMENT_METHOD,
      details: payload as unknown as Record<string, unknown>,
    },
  };

  return {
    payment_mandate_contents: updatedContents,
    user_authorization: userAuthorization,
  };
}

/**
 * Extract x402 PaymentPayload from a PaymentMandate.
 *
 * @param mandate - The PaymentMandate to extract the payload from
 * @returns The extracted x402 PaymentPayload, or undefined if not found
 */
export function extractX402Payload(mandate: PaymentMandate): PaymentPayload | undefined {
  const response = mandate.payment_mandate_contents.payment_response;
  if (response.method_name !== X402_PAYMENT_METHOD) return undefined;
  return response.details as unknown as PaymentPayload | undefined;
}

// ============================================================================
// Extension Helper
// ============================================================================

/**
 * Create an AP2 extension declaration for agent cards.
 *
 * @param roles - The AP2 roles this agent supports
 * @param required - Whether the extension is required
 * @returns An A2A extension declaration for AP2
 */
export function createAP2Extension(
  roles: AP2Role[] = ["merchant"],
  required: boolean = false,
): A2AExtension {
  return {
    uri: AP2_EXTENSION_URI,
    description: `AP2 payment agent (roles: ${roles.join(", ")}).`,
    required,
  };
}

// ============================================================================
// AgentCard & Header Helpers
// ============================================================================

/**
 * Create a complete payment extensions array for an AgentCard.
 * Returns [t402, x402, ap2?] extensions ready for capabilities.extensions.
 *
 * @param options - Configuration for extension declarations
 * @param options.ap2Roles - AP2 roles to advertise (omit to exclude AP2 extension)
 * @param options.t402Required - Whether the t402 extension is required
 * @param options.x402Required - Whether the x402 extension is required
 * @param options.ap2Required - Whether the AP2 extension is required
 * @returns Array of A2A extension declarations
 */
export function createPaymentExtensions(
  options: {
    ap2Roles?: AP2Role[];
    t402Required?: boolean;
    x402Required?: boolean;
    ap2Required?: boolean;
  } = {},
): A2AExtension[] {
  const extensions: A2AExtension[] = [
    createT402Extension(options.t402Required ?? false),
    createX402Extension(options.x402Required ?? false),
  ];
  if (options.ap2Roles) {
    extensions.push(createAP2Extension(options.ap2Roles, options.ap2Required ?? false));
  }
  return extensions;
}

/**
 * Get HTTP headers for A2A payment extension activation.
 * Returns the X-A2A-Extensions header with the x402 v0.2 URI.
 *
 * @param includeAP2 - Whether to include the AP2 extension URI
 * @returns Header name/value map
 */
export function getPaymentExtensionHeaders(includeAP2: boolean = false): Record<string, string> {
  const uris = [X402_A2A_EXTENSION_URI];
  if (includeAP2) {
    uris.push(AP2_EXTENSION_URI);
  }
  return { [A2A_EXTENSIONS_HEADER]: uris.join(", ") };
}

// ============================================================================
// DataPart Envelope Helpers
// ============================================================================

/**
 * Create a DataPart containing a CartMandate.
 *
 * @param cartMandate - The CartMandate to wrap
 * @returns An A2ADataPart containing the CartMandate
 */
export function createCartMandateDataPart(cartMandate: CartMandate): A2ADataPart {
  return {
    kind: "data",
    data: { [AP2_DATA_KEYS.CART_MANDATE]: cartMandate as unknown as Record<string, unknown> },
  };
}

/**
 * Create a DataPart containing a PaymentMandate.
 *
 * @param paymentMandate - The PaymentMandate to wrap
 * @returns An A2ADataPart containing the PaymentMandate
 */
export function createPaymentMandateDataPart(paymentMandate: PaymentMandate): A2ADataPart {
  return {
    kind: "data",
    data: { [AP2_DATA_KEYS.PAYMENT_MANDATE]: paymentMandate as unknown as Record<string, unknown> },
  };
}

/**
 * Create a DataPart containing an IntentMandate.
 *
 * @param intentMandate - The IntentMandate to wrap
 * @returns An A2ADataPart containing the IntentMandate
 */
export function createIntentMandateDataPart(intentMandate: IntentMandate): A2ADataPart {
  return {
    kind: "data",
    data: { [AP2_DATA_KEYS.INTENT_MANDATE]: intentMandate as unknown as Record<string, unknown> },
  };
}

/**
 * Create a DataPart containing a PaymentReceipt.
 *
 * @param receipt - The PaymentReceipt to wrap
 * @returns An A2ADataPart containing the PaymentReceipt
 */
export function createPaymentReceiptDataPart(receipt: PaymentReceipt): A2ADataPart {
  return {
    kind: "data",
    data: { [AP2_DATA_KEYS.PAYMENT_RECEIPT]: receipt as unknown as Record<string, unknown> },
  };
}

/**
 * Extract a CartMandate from an Artifact's parts.
 *
 * @param artifact - The A2A artifact to search for a CartMandate
 * @returns The extracted CartMandate, or undefined if not found
 */
export function extractCartMandateFromArtifact(artifact: A2AArtifact): CartMandate | undefined {
  if (!artifact.parts) return undefined;
  for (const part of artifact.parts) {
    if (part.kind === "data" && (part as A2ADataPart).data?.[AP2_DATA_KEYS.CART_MANDATE]) {
      return (part as A2ADataPart).data[AP2_DATA_KEYS.CART_MANDATE] as unknown as CartMandate;
    }
  }
  return undefined;
}

/**
 * Extract a PaymentMandate from a Message's parts.
 *
 * @param message - The A2A message to search for a PaymentMandate
 * @returns The extracted PaymentMandate, or undefined if not found
 */
export function extractPaymentMandateFromMessage(message: A2AMessage): PaymentMandate | undefined {
  for (const part of message.parts) {
    if (part.kind === "data" && (part as A2ADataPart).data?.[AP2_DATA_KEYS.PAYMENT_MANDATE]) {
      return (part as A2ADataPart).data[AP2_DATA_KEYS.PAYMENT_MANDATE] as unknown as PaymentMandate;
    }
  }
  return undefined;
}
